agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -0,0 +1,233 @@
+        return webSocket.waitForMessageOfType<ReceiveInviteMessage>(MESSAGE_TYPES.INVITE_GENERATED).invite
+    }
+
+    override fun waitForInvitedParty() {
+        if (getConnectionStatus() == AgentConnectionStatus.AGENT_CONNECTION_CONNECTED) {
+            webSocket.waitForMessageOfType<RequestReceivedMessage>(MESSAGE_TYPES.REQUEST_RECEIVED).also {
+                counterParty = IndyParty(it.did, it.endpoint)
+                sendJson(RequestSendResponseMessage(it.did))
+            }
+        } else {
+            throw AgentConnectionException("Agent is disconnected")
+        }
+    }
+
+    fun sendRequest(key: String) = webSocket.sendJson(SendRequestMessage(key))
+
+    override fun getCounterParty(): IndyParty? {
+        return counterParty
+    }
+
+    override fun sendCredentialOffer(offer: CredentialOffer) = webSocket.sendTypedMessage(offer, counterParty!!)
+
+    override fun receiveCredentialOffer(): CredentialOffer = webSocket.waitForTypedMessage()
+
+    override fun sendCredentialRequest(request: CredentialRequestInfo) = webSocket.sendTypedMessage(request, counterParty!!)
+
+    override fun receiveCredentialRequest(): CredentialRequestInfo = webSocket.waitForTypedMessage()
+
+    override fun sendCredential(credential: CredentialInfo) = webSocket.sendTypedMessage(credential, counterParty!!)
+
+    override fun receiveCredential(): CredentialInfo = webSocket.waitForTypedMessage()
+    override fun sendProofRequest(request: ProofRequest) = webSocket.sendTypedMessage(request, counterParty!!)
+
+    override fun receiveProofRequest(): ProofRequest = webSocket.waitForTypedMessage()
+
+    override fun sendProof(proof: ProofInfo) = webSocket.sendTypedMessage(proof, counterParty!!)
+
+    override fun receiveProof(): ProofInfo = webSocket.waitForTypedMessage()
+}
+
+class AgentWebSocketClient(serverUri: URI) : WebSocketClient(serverUri) {
+    private val log = KotlinLogging.logger {}
+
+    override fun onOpen(handshakedata: ServerHandshake?) {
+        log.info { "AgentConnection opened: $handshakedata" }
+    }
+
+    override fun onClose(code: Int, reason: String?, remote: Boolean) {
+        log.info { "AgentConnection closed: $code,$reason,$remote" }
+    }
+
+    val receivedMessages = mutableListOf<String>()
+
+    override fun onMessage(message: String?) {
+        log.info { "Message: $message" }
+        if (message != null)
+            synchronized(receivedMessages) { receivedMessages.add(message) }
+    }
+
+    override fun onError(ex: Exception?) {
+        log.warn(ex) { "AgentConnection error" }
+    }
+
+    fun sendJson(obj: Any) = send(SerializationUtils.anyToJSON(obj))
+
+    fun popMessageOfType(type: String): String? {
+        synchronized(receivedMessages) {
+            val result = receivedMessages.find { SerializationUtils.jSONToAny<Map<String, Any>>(it)["@type"].toString().contentEquals(type) }
```

Comment:
Does it possible to have intersection of agents' thread? 
E.g 2 clients and they both in the same phase of state machine. Should we care about sessionID (or other IDs) to handle which message exactly belongs to caller?

Reply:
It single client implementation, syncronization is done for interraction of websocket part with business logic, as they run in sepparate threads.
We do have correlationId for our messages, but no support for it in agent messages.

Reply:
Do you mean we have separated ws instance to cloud agent for each invite provided from our side?

Reply:
In current implementation we cant have multiple session at the same time, even if we create multiple client, only one WS session possible, so previous will be dropped.
We need to implement multi-session dispatcher on web-socket client layer.

Reply:
Need to create task in backlog with very high priority.

Reply:
Lets make it clear. Current agent implementation supports only one client attached. So does our client. Current client implementation can work only with one counter party. We   can add multiple counter party support if we add message dispatcher.

Reply:
https://github.com/hyperledger-labs/cordentity/projects/1#card-18735693